source/clear-linux/tutorials/fmv.rst

```diff
@@ -0,0 +1,268 @@
+
+The make-fmv-patch.pl take two arguments: <buildlog> and <sourcecode>. Replace with the proper values and execute:
+
+.. code-block:: bash
+
+        perl make-fmv-patch.pl <buildlog> <sourcecode>
+
+The command generates the following :file:`example.c.patch` patch:
+
+.. code-block:: console
+
+    --- ./example.c 2017-09-27 16:05:42.279505430 +0000
+    +++ ./example.c~    2017-09-27 16:19:11.691544026 +0000
+    @@ -5,6 +5,7 @@
+
+     int a[256], b[256], c[256];
+
+    +__attribute__((target_clones("avx2","arch=atom","default")))
+     void foo(){
+         int i,x;
+         for (x=0; x<MAX; x++){
+
+The `make-fmv-patch` is recommended to add the attribute generating the
+target clones on the function foo. Thus, we can have the following code:
+
+.. code-block:: c
+
+    #include <stdio.h>
+    #include <stdlib.h>
+    #include <sys/time.h>
+    #define MAX 1000000
+
+    int a[256], b[256], c[256];
+
+    __attribute__((target_clones("avx2","arch=atom","default")))
+    void foo(){
+        int i,x;
+        for (x=0; x<MAX; x++){
+            for (i=0; i<256; i++){
+                a[i] = b[i] + c[i];
+            }
+        }
+    }
+
+
+    int main(){
+        foo();
+        return 0;
+    }
+
+By changing the value of the `$avx2` variable, we can change the target clones when adding the patches or in the make-fmv-patch.pl script:
+
+.. code-block:: perl
+
+    my $avx2 = '__attribute__((target_clones("avx2","arch=atom","default")))'."\n";
+
+Compile the code again with FMV and add the option to analyze the `objdump`:
+
+.. code-block:: bash
+
+    gcc -O3 example.c -o example -g
+    objdump -S example | less
+
+You can see the multiple clones of the foo function:
+
+.. code-block:: console
+
+    foo
+    foo.avx2.0
+    foo.arch_atom.1
+
+The cloned functions use AVX2 registers and vectorized instructions. You can verify this with:
+
+.. code-block:: assembly
+
+    vpaddd (%r8,%rax,1),%ymm0,%ymm0
+    vmovdqu %ymm0,(%rcx,%rax,1)
+
+FTT project example
+*******************
+
+To follow the same approach with a package like FFT, we must get the build log file with the `-fopt-info-vec` flag:
+
+::
+
+    ~/make-fmv-patch/make-fmv-patch.pl results/build.log fftw-3.3.6-pl2/
+
+    patching fftw-3.3.6-pl2/libbench2/verify-lib.c @ lines (36 114 151 162 173 195 215 284)
+    patching fftw-3.3.6-pl2/tools/fftw-wisdom.c @ lines (150)
+    patching fftw-3.3.6-pl2/libbench2/speed.c @ lines (26)
+    patching fftw-3.3.6-pl2/tests/bench.c @ lines (27)
+    patching fftw-3.3.6-pl2/libbench2/util.c @ lines (181)
+    patching fftw-3.3.6-pl2/libbench2/problem.c @ lines (229)
+    patching fftw-3.3.6-pl2/tests/fftw-bench.c @ lines (101 147 162 249)
+    patching fftw-3.3.6-pl2/libbench2/mp.c @ lines (79 190 215)
+    patching fftw-3.3.6-pl2/libbench2/caset.c @ lines (5)
+    patching fftw-3.3.6-pl2/libbench2/verify-r2r.c @ lines (44 187 197 207 316 333 723)
+
+Thus, files like :file:`fftw-3.3.6-pl2/tools/fftw-wisdom.c.patch` generate
```

Comment:
Please include further description to "files like". Perhaps we can also replace "like" with "similar to" in the second instance of the word.

Reply:
I disagree. We do not need to replace 'like' with anything else. In technical writing we value using the same word for the same meaning. That's one key difference with other forms of writing.

Reply:
Repetition of a word is fine. I take issue with the word "like" as it provokes the question: what constitutes y to be like x. In this case, what makes a file like `fftw-3.3.6-pl2/tools/fftw-wisdom.c.patch`? I'd like that question to be answered in lieu of the "like". Maybe it is: files in x directory. I thought the second "like" was more straight forward as I assumed it stood for "with the following structure".

Reply:
In this instance, fftw-3.3.6-pl2/tools/fftw-wisdom.c.patch is an example patch file. My resistance here is due to clarity. As it is written, the text is very clear: If you feed A to the script you get B out. The word "like" is just further emphasizing the exemplary nature of the file and the output. I fear that by adding a lengthy explanation about the inner structure of the file and output we will obscure this clear example without providing truly valuable information. This tutorial is different than others we made since the results are less tangible. At the end of the tutorial, you don't get an application running but optimized code for a specific purpose. The examples are just that, examples of what code looks like before and after using FMV. Thus, the readers do not need to know all the details about the file to understand what FMV does since the readers will use their own files with their own code.